File: app/src/main/java/com/example/digitalmonk/service/accessibility/detectors/ShortsDetector.java
package com.example.digitalmonk.service.accessibility.detectors;

import android.view.accessibility.AccessibilityNodeInfo;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ShortsDetector {

    private static final Set<String> BLOCKED_PACKAGES = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
            "com.ss.android.ugc.trill",
            "com.zhiliaoapp.musically",
            "com.ss.android.ugc.aweme"
    )));

    private static final Set<String> SHORTS_VIEW_IDS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
            "com.google.android.youtube:id/reel_recycler",        // YouTube Shorts
            "app.revanced.android.youtube:id/reel_recycler",      // ReVanced YouTube
            "com.instagram.android:id/root_clips_layout",         // Instagram Reels
            "com.instagram.android:id/reply_bar_container"        // Instagram Inbox Reels
    )));

    /**
     * Private constructor for Utility Class.
     */
    private ShortsDetector() {}

    /**
     * Determines if the current screen contains short-form video content.
     */
    public static boolean shouldBlock(AccessibilityNodeInfo rootNode, String packageName) {
        if (packageName == null) return false;

        // 1. Check if the entire app is blocked (e.g., TikTok)
        if (BLOCKED_PACKAGES.contains(packageName)) return true;

        // 2. Check if a specific "Shorts" View ID is on the screen
        if (rootNode == null) return false;

        for (String viewId : SHORTS_VIEW_IDS) {
            if (hasViewId(rootNode, viewId)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Searches the Android UI hierarchy for a specific View ID.
     */
    private static boolean hasViewId(AccessibilityNodeInfo root, String viewId) {
        try {
            List<AccessibilityNodeInfo> nodes = root.findAccessibilityNodeInfosByViewId(viewId);
            return nodes != null && !nodes.isEmpty();
        } catch (Exception e) {
            return false;
        }
    }
}

File: app/src/main/java/com/example/digitalmonk/service/accessibility/detectors/ShortsDetector.kt
package com.example.digitalmonk.service.accessibility.detectors

import android.view.accessibility.AccessibilityNodeInfo

/**
 * Detects whether the current screen is showing short-form video content.
 *
 * Detection strategy:
 *  1. Package-level block — entire app is short-form (e.g. TikTok).
 *  2. View-ID detection — Shorts/Reels player embedded inside a larger app
 *     (e.g. YouTube Shorts inside the YouTube app).
 *
 * To add support for a new app: append to [BLOCKED_PACKAGES] or [SHORTS_VIEW_IDS].
 */
object ShortsDetector {

    // ── Whole-app blocks ──────────────────────────────────────────────────────
    private val BLOCKED_PACKAGES = setOf(
        "com.ss.android.ugc.trill",       // TikTok (some regions)
        "com.zhiliaoapp.musically",        // TikTok (global)
        "com.ss.android.ugc.aweme"         // TikTok (CN / Douyin)
    )

    // ── View-ID detection ─────────────────────────────────────────────────────
    private val SHORTS_VIEW_IDS = setOf(
        "com.google.android.youtube:id/reel_recycler",        // YouTube Shorts
        "app.revanced.android.youtube:id/reel_recycler",      // ReVanced YouTube
        "com.instagram.android:id/root_clips_layout",          // Instagram Reels
        "com.instagram.android:id/reply_bar_container"         // Instagram Inbox Reels
    )

    fun shouldBlock(rootNode: AccessibilityNodeInfo?, packageName: String?): Boolean {
        if (packageName == null) return false
        if (packageName in BLOCKED_PACKAGES) return true
        if (rootNode == null) return false
        return SHORTS_VIEW_IDS.any { viewId -> hasViewId(rootNode, viewId) }
    }

    private fun hasViewId(root: AccessibilityNodeInfo, viewId: String): Boolean =
        try {
            val nodes = root.findAccessibilityNodeInfosByViewId(viewId)
            nodes != null && nodes.isNotEmpty()
        } catch (e: Exception) {
            false
        }
}